libraries/chain/backing_store/tests/benchmark_kv.cpp

```diff
@@ -0,0 +1,337 @@
+      for (auto& key: work_set) {
+         kv_context_ptr->kv_set(contract, key.c_str(), key.size(), value.c_str(), value.size(), payer);
+      }
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_it_next(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
+
+   uint32_t found_key_size, found_value_size;
+   auto it = kv_context_ptr->kv_it_create(contract, "", 0);
+   it->kv_it_next(&found_key_size, &found_value_size);
+
+   while (it->kv_it_status() != kv_it_stat::iterator_end) {
+      it->kv_it_next(&found_key_size, &found_value_size);
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_it_key_value(cmd_args& args, const std::unique_ptr<kv_context> kv_context_ptr, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   uint32_t offset = 0;
+   char* dest = new char[args.value_size];
+   uint32_t actual_size;
+   uint32_t found_key_size, found_value_size;
+
+   start = std::chrono::steady_clock::now();
+   getrusage(RUSAGE_SELF, &usage_start);
+
+   auto it = kv_context_ptr->kv_it_create(contract, "", 0);
+   it->kv_it_next(&found_key_size, &found_value_size);
+
+   while (it->kv_it_status() != kv_it_stat::iterator_end) {
+      it->kv_it_key(offset, dest, found_key_size, actual_size);
+      it->kv_it_value(offset, dest, found_value_size, actual_size);
+
+      it->kv_it_next(&found_key_size, &found_value_size);
+   }
+
+   getrusage(RUSAGE_SELF, &usage_end);
+   end = std::chrono::steady_clock::now();
+}
+
+void benchmark_operation(cmd_args& args, std::unique_ptr<kv_context> kv_context_ptr, wall_clock_t& start, wall_clock_t& end, rusage& usage_start, rusage& usage_end) {
+   create_key_values(args, kv_context_ptr);
+   std::vector<std::string> work_set = get_work_set_keys(args);
+
+   if (args.operation == "get" ) {
+      benchmark_get(args, std::move(kv_context_ptr), work_set, start, end, usage_start, usage_end);
+   } else if (args.operation == "get_data" ) {
+      benchmark_get_data(args, std::move(kv_context_ptr), work_set, start, end, usage_start, usage_end);
+   } else if (args.operation == "set" ) {
+      benchmark_set(args, std::move(kv_context_ptr), work_set, start, end, usage_start, usage_end);
+   } else if (args.operation == "it_next" ) {
+      benchmark_it_next(args, std::move(kv_context_ptr), start, end, usage_start, usage_end);
+   } else if (args.operation == "it_key_value" ) {
+      benchmark_it_key_value(args, std::move(kv_context_ptr), start, end, usage_start, usage_end);
+   };
+}
+
+void print_results(cmd_args& args, wall_clock_t& start, wall_clock_t& end, const rusage& usage_start, const rusage& usage_end) {
+   std::string title = "Benchmarked \"" + args.operation + "\" for ";
```

Comment:
Maybe consider `stringstream` instead.

Reply:
stringstream performance is terrible, not that it matters here. I much rather see std::string and to_string instead of stringstream.

Reply:
I would like to see the metrics on that.  Assuming that you don't know the ultimate size of your string to be able to reserve the space before hand, appending to a string would result in reallocations and copies.

Reply:
Google it. stringstream you are paying for all the overhead of locale and the terrible iostream library.

Reply:
I prefer to std::string too for simplicity.

Reply:
Since it's all just going to `cout` below, is a temporary string needed at all?

Reply:
I have removed the strings.